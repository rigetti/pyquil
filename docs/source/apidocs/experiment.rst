Experiment
==========

The ``experiment`` module offers a schema and utilities for succinctly expressing commonly
used applications and algorithms in near-term quantum programming. A ``TomographyExperiment``
is intended to be consumed by the ``QuantumComputer.experiment`` method.

**NOTE**: When working with the `experiment` method, the following declared memory labels are
reserved:

 - "preparation_alpha", "preparation_beta", and "preparation_gamma"
 - "measurement_alpha", "measurement_beta", and "measurement_gamma"
 - "symmetrization"
 - "ro"

.. currentmodule:: pyquil.experiment

Schema
------

.. autoclass:: pyquil.experiment.TomographyExperiment

   .. rubric:: Methods

   .. autosummary::
        :toctree: autogen
        :template: autosumm.rst

        ~TomographyExperiment.get_meas_qubits
        ~TomographyExperiment.get_meas_registers
        ~TomographyExperiment.generate_experiment_program
        ~TomographyExperiment.build_setting_memory_map
        ~TomographyExperiment.build_symmetrization_memory_maps
        ~TomographyExperiment.generate_calibration_experiment

.. autoclass:: pyquil.experiment.ExperimentSetting

.. autoclass:: pyquil.experiment.ExperimentResult

.. autoclass:: pyquil.experiment.SymmetrizationLevel

.. autoclass:: pyquil.experiment.CalibrationMethod

Utilities
---------

.. autofunction:: pyquil.experiment.bitstrings_to_expectations
.. autofunction:: pyquil.experiment.correct_experiment_result
.. autofunction:: pyquil.experiment.merge_memory_map_lists
.. autofunction:: pyquil.experiment.ratio_variance
.. autofunction:: pyquil.experiment.read_json
.. autofunction:: pyquil.experiment.to_json
