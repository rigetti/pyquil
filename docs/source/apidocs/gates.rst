Gates and Instructions
======================

A :py:class:`~pyquil.quil.Program` is effectively a list of gates and instructions which
can be created using the function documented in this section::

    >>> p = Program()
    >>> p += H(qubit=0)
    >>> p += RY(angle=pi/3, qubit=1)
    >>> p += CNOT(0, 1)
        ...


.. currentmodule:: pyquil.gates

Native gates for Rigetti QPUs
-----------------------------

Physical quantum processors can enact a subset of all named gates. Luckily,
a small set of gates is universal for quantum computation, so all named gates
can be enacted by suitable combinations of physically realizable gates. Rigetti's
superconducting quantum processors can perform :py:func:`RX` with ``angle=+-pi/2`` or
``angle=+-pi``, :py:func:`RZ` with an arbitrary angle, and :py:func:`CZ` interactions
between neighboring qubits. Rigetti QPUs can natively measure in the computational (Z) basis.

.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    RX
    RZ
    CZ
    MEASURE


All gates and instructions
--------------------------

In general, you will write a quantum program using the full suite of Quil gates and instructions
and use :ref:`the Quil compiler <compiler>` to compile programs into the target instruction
set (ISA). The full list of quantum gates and classical Quil instructions is enumerated here.

.. rubric:: Single-qubit gates
.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    I
    X
    Y
    Z
    H
    S
    T
    RX
    RY
    RZ
    PHASE

.. rubric:: Multi-qubit gates
.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    CZ
    CNOT
    CCNOT
    CPHASE00
    CPHASE01
    CPHASE10
    CPHASE
    SWAP
    CSWAP
    ISWAP
    PSWAP


.. rubric:: Classical instructions
.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    WAIT
    RESET
    NOP
    HALT
    MEASURE
    NEG
    NOT
    AND
    OR
    IOR
    XOR
    MOVE
    EXCHANGE
    LOAD
    STORE
    CONVERT
    ADD
    SUB
    MUL
    DIV
    EQ
    LT
    LE
    GT
    GE

.. rubric:: Collections
.. autosummary::

    QUANTUM_GATES
    STANDARD_INSTRUCTIONS

