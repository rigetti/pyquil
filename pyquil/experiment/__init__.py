__all__ = [
    "bitstrings_to_expectations",
    "CalibrationMethod",
    "correct_experiment_result",
    "Experiment",
    "ExperimentResult",
    "ExperimentSetting",
    "get_results_by_qubit_groups",
    "group_settings",
    "merge_disjoint_experiments",
    "merge_memory_map_lists",
    "minusX",
    "minusY",
    "minusZ",
    "OperatorEncoder",
    "plusX",
    "plusY",
    "plusZ",
    "ratio_variance",
    "read_json",
    "SIC0",
    "SIC1",
    "SIC2",
    "SIC3",
    "SymmetrizationLevel",
    "TensorProductState",
    "to_json",
    "zeros_state",
]

from pyquil.experiment._calibration import CalibrationMethod
from pyquil.experiment._group import (
    merge_disjoint_experiments,
    get_results_by_qubit_groups,
    group_settings,
)
from pyquil.experiment._main import (
    Experiment,
    OperatorEncoder,
    read_json,
    to_json,
)
from pyquil.experiment._memory import merge_memory_map_lists
from pyquil.experiment._result import (
    ExperimentResult,
    bitstrings_to_expectations,
    correct_experiment_result,
    ratio_variance,
)
from pyquil.experiment._setting import (
    _OneQState,
    ExperimentSetting,
    SIC0,
    SIC1,
    SIC2,
    SIC3,
    TensorProductState,
    minusX,
    minusY,
    minusZ,
    plusX,
    plusY,
    plusZ,
    zeros_state,
)
from pyquil.experiment._symmetrization import SymmetrizationLevel
