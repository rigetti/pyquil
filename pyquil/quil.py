#!/usr/bin/python
##############################################################################
# Copyright 2016-2017 Rigetti Computing
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
##############################################################################
"""
Module for creating and defining Quil programs.
"""

from .quilbase import (InstructionGroup,
                       Addr,
                       While,
                       If,
                       DefGate,
                       Gate,
                       Measurement,
                       merge_resource_managers)

from .gates import MEASURE, STANDARD_GATES
from math import pi


class Program(InstructionGroup):
    def __init__(self, *instructions):
        super(Program, self).__init__()
        self.inst(*instructions)
        self.defined_gates = []

    def synthesize(self, resource_manager=None):
        self.resource_manager.reset()
        return super(Program, self).synthesize(resource_manager)

    def __add__(self, other):
        """
        Concatenate two programs together, returning a new one.

        :param Program other: Another program or instruction to concatenate to this one.
        :return: A newly concatenated program.
        :rtype: Program
        """
        if isinstance(other, Program):
            p = Program()
            p.defined_gates = self.defined_gates + other.defined_gates
            p.actions = self.actions + other.actions
            p.resource_manager = merge_resource_managers(self.resource_manager,
                                                         other.resource_manager)
            return p
        else:
            return super(Program, self).__add__(other)

    def __getitem__(self, index):
        """
        Allows indexing into the program to get an action.
        :param index: The action at the specified index.
        :return:
        """
        return self.actions[index]

    def __iter__(self):
        """
        Allow built in iteration through a program's actions, e.g. [a for a in Program(X(0))]
        :return:
        """
        return self.actions.__iter__()

    def get_qubits(self):
        """
        :return: a set of all the qubit indices allocated in this program, synthesizing freely
        allocated qubits if neccessary.
        :rtype: set
        """
        allocated_qubits = set()
        self.synthesize()
        for ii, action in self:
            if isinstance(action, Gate):
                qubit_indices = {qq.index() for qq in action.arguments}
            elif isinstance(action, Measurement):
                qubit_indices = {action.arguments[0].index()}
            else:
                continue
            allocated_qubits = set.union(allocated_qubits, qubit_indices)
        return allocated_qubits

    def defgate(self, name, matrix):
        """
        Define a new static gate.

        :param string name: The name of the gate.
        :param array-like matrix: List of lists or Numpy 2d array.
        :return: The Program instance.
        :rtype: Program
        """
        self.defined_gates.append(DefGate(name, matrix))
        return self

    def measure(self, qubit_index, classical_reg):
        """
        Measures a qubit at qubit_index and puts the result in classical_reg

        :param int qubit_index: The address of the qubit to measure.
        :param int classical_reg: The address of the classical bit to store the result.
        :returns: The Quil Program with the appropriate measure instruction appended, e.g.
                  MEASURE 0 [1]
        :rtype: Program
        """
        return self.inst(MEASURE(qubit_index, classical_reg))

    def measure_all(self, *qubit_reg_pairs):
        """
        Measures many qubits into their specified classical bits, in the order
        they were entered.

        :param Tuple qubit_reg_pairs: Tuples of qubit indices paired with classical bits.
        :return: The Quil Program with the appropriate measure instructions appended, e.g.
                  MEASURE 0 [1]
                  MEASURE 1 [2]
                  MEASURE 2 [3]
        :rtype: Program
        """
        for qubit_index, classical_reg in qubit_reg_pairs:
            self.inst(MEASURE(qubit_index, classical_reg))
        return self

    def while_do(self, classical_reg, q_program):
        """
        While a classical register at index classical_reg is 1, loop q_program

        :param int classical_reg: The classical register to check
        :param Program q_program: The Quil program to loop.
        :return: The Quil Program with the loop instructions added.
        :rtype: Program
        """
        w_loop = While(Addr(classical_reg))
        w_loop.Body.inst(q_program)
        return self.inst(w_loop)

    def if_then(self, classical_reg, if_program, else_program=None):
        """
        If the classical register at index classical reg is 1, run if_program, else run
        else_program.

        :param int classical_reg: The classical register to check as the condition
        :param Program if_program: A Quil program to execute if classical_reg is 1
        :param Program else_program: A Quil program to execute if classical_reg is 0. This
            argument is optional and defaults to an empty Program.
        :returns: The Quil Program with the branching instructions added.
        :rtype: Program
        """

        else_program = else_program if else_program is not None else Program()

        branch = If(Addr(classical_reg))
        branch.Then.inst(if_program)
        branch.Else.inst(else_program)
        return self.inst(branch)

    def dagger(self, inv_dict=None, suffix="-INV"):
        """
        Creates the conjugate transpose of the Quil program. The program must not
        contain any irreversible actions (measurement, control flow, qubit allocation).

        :return: The Quil program's inverse
        :rtype: Program

        """

        for action in self.actions:
            assert action[0] == 0, "Program must be valid Protoquil"
            gate = action[1]
            assert not isinstance(gate, Measurement), "Program cannot contain measurements"
            assert not isinstance(gate, While) and not isinstance(gate, If), \
                "Program cannot contain control flow"

        daggered = Program()

        for gate in self.defined_gates:
            if inv_dict == None or gate.name not in inv_dict:
                daggered.defgate(gate.name + suffix, gate.matrix.T.conj())

        for action in self.actions[::-1]:
            gate = action[1]
            if gate.operator_name in STANDARD_GATES:
                if gate.operator_name == "S":
                    daggered.inst(STANDARD_GATES["PHASE"](-pi / 2, *gate.arguments))
                elif gate.operator_name == "T":
                    daggered.inst(STANDARD_GATES["RZ"](pi / 4, *gate.arguments))
                elif gate.operator_name == "ISWAP":
                    daggered.inst(STANDARD_GATES["PSWAP"](pi / 2, *gate.arguments))
                else:
                    negated_params = list(map(lambda x : -1*x, gate.parameters))
                    daggered.inst(STANDARD_GATES[gate.operator_name](*(negated_params+gate.arguments)))
            else:
                if inv_dict == None or gate.operator_name not in inv_dict:
                    gate_inv_name = gate.operator_name + suffix
                else:
                    gate_inv_name = inv_dict[gate.operator_name]

                daggered.inst(tuple([gate_inv_name] + gate.arguments))

        return daggered

    def out(self):
        """
        Converts the Quil program to a readable string.

        :return: String form of a program
        :rtype: string
        """
        s = ""
        for dg in self.defined_gates:
            s += dg.out()
            s += "\n"
        s += super(Program, self).out()
        return s

def merge_programs(prog_list):
    """
    Merges a list of pyQuil programs into a single one by appending them in sequence

    :param list prog_list: A list of pyquil programs
    :return: a single pyQuil program
    :rtype: Program
    """
    return sum(prog_list, Program())
