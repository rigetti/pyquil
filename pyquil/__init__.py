from pyquil._version import pyquil_version
from pyquil.quil import Program
from pyquil.api import list_quantum_computers, get_qc

__version__ = pyquil_version
