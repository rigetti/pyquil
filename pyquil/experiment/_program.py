##############################################################################
# Copyright 2016-2019 Rigetti Computing
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
##############################################################################

from typing import Sequence

import numpy as np

from pyquil.gates import MEASURE, RX, RZ
from pyquil.quil import Program


def parameterized_euler_rotations(
    qubits: Sequence[int],
    *,
    prefix: str,
    suffix_alpha: str = "alpha",
    suffix_beta: str = "beta",
    suffix_gamma: str = "gamma",
) -> Program:
    """
    Given a number of qubits (n), build a ``Program`` containing a ZXZXZ-decomposed gate on each
    qubit, where each ``RZ`` is parameterized by declared values with labels given by the "prefix"
    and "suffix" arguments. Put more plainly, the resulting Quil program on n qubits is::

        RZ(alpha_label[0]) 0
        RX(pi/2) 0
        RZ(beta_label[0]) 0
        RX(-pi/2) 0
        RZ(gamma_label[0]) 0
        ...
        RZ(alpha_label[n-1]) n-1
        RX(pi/2) n-1
        RZ(beta_label[0]) n-1
        RX(-pi/2) n-1
        RZ(gamma_label[n-1]) n-1

    :param qubits: The number of qubits (n).
    :param prefix: The prefix for the declared memory region labels. For example, if the prefix
        is "preparation" and the alpha, beta, and gamma suffixes are left as default, the labels
        would be "preparation_alpha", "preparation_beta", and "preparation_gamma".
    :param suffix_alpha: The suffix for the "alpha" memory region label, which corresponds to the
        first (rightmost) ``Z`` in the ZXZXZ decomposition. Defaults to "alpha".
    :param suffix_beta: The suffix for the "beta" memory region label, which corresponds to the
        second (middle) ``Z`` in the ZXZXZ decomposition. Defaults to "beta".
    :param suffix_gamma: The suffix for the "gamma" memory region label, which corresponds to the
        last (leftmost) ``Z`` in the ZXZXZ decomposition. Defaults to "gamma".
    :return: A ``Program`` containing a 3 parameterized ``RZ``s and 2 fixed ``RX``s per qubit.
    """
    alpha_label = f"{prefix}_{suffix_alpha}"
    beta_label = f"{prefix}_{suffix_beta}"
    gamma_label = f"{prefix}_{suffix_gamma}"

    p = Program()

    alpha = p.declare(alpha_label, "REAL", len(qubits))
    beta = p.declare(beta_label, "REAL", len(qubits))
    gamma = p.declare(gamma_label, "REAL", len(qubits))

    for idx, q in enumerate(qubits):
        p += RZ(alpha[idx], q)
        p += RX(np.pi / 2, q)
        p += RZ(beta[idx], q)
        p += RX(-np.pi / 2, q)
        p += RZ(gamma[idx], q)

    return p


def parameterized_single_qubit_state_preparation(
    qubits: Sequence[int], label: str = "preparation"
) -> Program:
    """
    Given a number of qubits, produce a program as in ``parameterized_euler_rotations`` where each
    memory region is prefixed by ``label``, where label defaults to "preparation".

    :param qubits: The number of qubits (n).
    :param label: The prefix to use when declaring memory in ``parameterized_euler_rotations``.
    :return: A parameterized ``Program`` that can be used to prepare a product state.
    """
    return parameterized_euler_rotations(qubits, prefix=label)


def parameterized_single_qubit_measurement_basis(
    qubits: Sequence[int], label: str = "measurement"
) -> Program:
    """
    Given a number of qubits, produce a program as in ``parameterized_euler_rotations`` where each
    memory region is prefixed by ``label``, where label defaults to "measurement".

    :param qubits: The number of qubits (n).
    :param label: The prefix to use when declaring memory in ``parameterized_euler_rotations``.
    :return: A parameterized ``Program`` that can be used to measure in a multi-qubit Pauli basis.
    """
    return parameterized_euler_rotations(qubits, prefix=label)


def parameterized_readout_symmetrization(
    qubits: Sequence[int], label: str = "symmetrization"
) -> Program:
    """
    Given a number of qubits (n), produce a ``Program`` with an ``RX`` instruction on qubits
    0 through n-1, parameterized by memory regions label[0] through label[n-1], where "label"
    defaults to "symmetrization".

    :param qubits: The number of qubits (n).
    :param label: The name of the declared memory region.
    :return: A ``Program`` with parameterized ``RX`` gates on n qubits.
    """
    p = Program()
    symmetrization = p.declare(f"{label}", "REAL", len(qubits))
    for idx, q in enumerate(qubits):
        p += RX(symmetrization[idx], q)
    return p


def measure_qubits(qubits: Sequence[int]) -> Program:
    """
    Given a number of qubits (n), produce a ``Program`` with a ``MEASURE`` instruction on qubits
    0 through n-1, with corresponding readout registers ro[0] through ro[n-1].

    :param qubits: The number of qubits (n).
    :return: A ``Program`` that measures n qubits.
    """
    p = Program()
    ro = p.declare("ro", "BIT", len(qubits))
    for idx, q in enumerate(qubits):
        p += MEASURE(q, ro[idx])
    return p
