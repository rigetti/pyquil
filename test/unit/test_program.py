import numpy as np
from syrupy.assertion import SnapshotAssertion

from pyquil import Program
from pyquil.quil import AbstractInstruction, Declare, Measurement
from pyquil.quilbase import Reset, ResetQubit, Delay, DelayFrames, DelayQubits, Frame
from pyquil.quilatom import Label, MemoryReference
from pyquil.experiment._program import (
    measure_qubits,
    parameterized_single_qubit_measurement_basis,
    parameterized_single_qubit_state_preparation,
    parameterized_readout_symmetrization,
)
from pyquil.gates import MEASURE, RX, RZ, H


def test_measure_qubits():
    p = Program()
    ro = p.declare("ro", "BIT", 2)
    p += MEASURE(0, ro[0])
    p += MEASURE(1, ro[1])
    assert measure_qubits([0, 1]).out() == p.out()


def test_parameterized_single_qubit_measurement_basis():
    p = Program()
    alpha = p.declare("measurement_alpha", "REAL", 2)
    beta = p.declare("measurement_beta", "REAL", 2)
    gamma = p.declare("measurement_gamma", "REAL", 2)
    for idx, q in enumerate(range(2)):
        p += RZ(alpha[idx], q)
        p += RX(np.pi / 2, q)
        p += RZ(beta[idx], q)
        p += RX(-np.pi / 2, q)
        p += RZ(gamma[idx], q)
    assert parameterized_single_qubit_measurement_basis([0, 1]).out() == p.out()


def test_parameterized_single_qubit_state_preparation():
    p = Program()
    alpha = p.declare("preparation_alpha", "REAL", 2)
    beta = p.declare("preparation_beta", "REAL", 2)
    gamma = p.declare("preparation_gamma", "REAL", 2)
    p += RZ(alpha[0], 0)
    p += RX(np.pi / 2, 0)
    p += RZ(beta[0], 0)
    p += RX(-np.pi / 2, 0)
    p += RZ(gamma[0], 0)
    p += RZ(alpha[1], 1)
    p += RX(np.pi / 2, 1)
    p += RZ(beta[1], 1)
    p += RX(-np.pi / 2, 1)
    p += RZ(gamma[1], 1)
    assert parameterized_single_qubit_state_preparation([0, 1]).out() == p.out()


def test_parameterized_readout_symmetrization():
    p = Program()
    symmetrization = p.declare("symmetrization", "REAL", 2)
    p += RX(symmetrization[0], 0)
    p += RX(symmetrization[1], 1)
    assert parameterized_readout_symmetrization([0, 1]).out() == p.out()


def test_adding_does_not_mutate():
    # https://github.com/rigetti/pyquil/issues/1476
    p1 = Program(
        """
DEFCAL RX(pi/2) 32:
    FENCE 32
    NONBLOCKING PULSE 32 "rf" drag_gaussian(duration: 3.2e-08, fwhm: 8e-09, t0: 1.6e-08, anh: -190000000.0, alpha: -1.8848698349348032, scale: 0.30631340170943533, phase: 0.0, detuning: 1622438.2425563578)
    FENCE 32

RX(pi/2) 32
"""
    )
    original_p1 = p1.copy()
    p2 = Program(
        """
DEFCAL RX(pi/2) 33:
    FENCE 33
    NONBLOCKING PULSE 33 "rf" drag_gaussian(duration: 2e-08, fwhm: 5e-09, t0: 1e-08, anh: -190000000.0, alpha: -0.9473497322033984, scale: 0.25680107985232403, phase: 0.0, detuning: 1322130.5458282642)
    FENCE 33

RX(pi/2) 33
"""
    )
    p_all = p1 + p2
    assert str(p1) == str(original_p1)
    assert p1.calibrations != p_all.calibrations


def test_with_loop(snapshot: SnapshotAssertion):
    p = Program(
        """DECLARE ro BIT
DECLARE shot_count INTEGER
MEASURE q ro
JUMP-UNLESS @end-reset ro
X q
LABEL @end-reset

DEFCAL I 0:
    DELAY 0 1.0
DEFFRAME 0 \"rx\":
    HARDWARE-OBJECT: \"hardware\"
DEFWAVEFORM custom:
    1,2
I 0
"""
    )
    p_copy = p.copy()

    looped = p.with_loop(100, MemoryReference("shot_count"), Label("start-loop"), Label("end-loop"))

    assert p_copy == p
    assert looped.out() == snapshot


def test_filter_program():
    program = Program(Declare("ro", "BIT", 1), MEASURE(0, MemoryReference("ro", 1)), H(0))

    def predicate(instruction: AbstractInstruction) -> bool:
        if isinstance(instruction, Declare):
            return instruction.name != "ro"
        elif isinstance(instruction, Measurement):
            return instruction.classical_reg.name != "ro"
        else:
            return True

    filtered_program = program.filter_instructions(predicate)
    assert filtered_program == Program(H(0))


def test_filter_quil_t():
    non_quil_t_program = Program(
        """DECLARE ro BIT[1]
H 0
CNOT 0 1
MEASURE 0 ro[0]
MEASURE 1 ro[0]
WAIT
"""
    )

    quil_t_program = Program(
        """
DEFCAL I q:
    DELAY q 4e-08
DEFFRAME 0 "rf":
    DIRECTION: "tx"
DEFCAL MEASURE 0 addr:
    FENCE 0
DEFWAVEFORM q44_q45_cphase/sqrtCPHASE:
    0.0, 0.0, 0.00027685415721916584
CAPTURE 10 "ro_rx" boxcar_kernel(duration: 1.6e-06, scale: 1.0, phase: 0.0, detuning: 0.0) q10_unclassified[0]
NONBLOCKING CAPTURE 10 "ro_rx" boxcar_kernel(duration: 1.6e-06, scale: 1.0, phase: 0.0, detuning: 0.0) q10_unclassified[0]
FENCE 0
PULSE 0 "rf_f12" gaussian(duration: 6.000000000000001e-08, fwhm: 1.5000000000000002e-08, t0: 3.0000000000000004e-08, scale: 0.16297407445283926, phase: 0.0, detuning: 0)
RAW-CAPTURE 0 "out" 200000000 iqs[0]
SET-FREQUENCY 0 "xy" 5400000000
SET-PHASE 0 "xy" pi
SET-SCALE 0 "xy" pi
SHIFT-FREQUENCY 0 "ro" 6100000000
SHIFT-PHASE 0 "xy" (-pi)
SHIFT-PHASE 0 "xy" (%theta*(2/pi))
SWAP-PHASES 2 3 "xy" 3 4 "xy";
"""
    )
    full_program = non_quil_t_program + quil_t_program
    assert full_program.remove_quil_t_instructions() == non_quil_t_program


def test_compatibility_layer():
    """
    Test that the compatibility layer that transforms pyQuil instructions to quil instructions works as intended.
    """
    # Note: `quil` re-orders some instructions in a program (e.g. by shuffling DECLAREs to the top). This isn't a
    # problem for the current set of instructions we're testing, but it's something to keep in mind if we add more.
    instructions = [
        ResetQubit(0),
        Reset(),
        Delay([Frame([0], "frame")], [0], 0.01),
        DelayFrames([Frame([], "frame")], 0.01),
        DelayQubits([0, 1], 0.01),
    ]
    program = Program(instructions)
    for (original, transformed) in zip(instructions, program):
        assert isinstance(transformed, AbstractInstruction)
        print("type(transformed):", type(transformed))
        print("type(original):", type(original))
        assert isinstance(transformed, type(original))
        assert transformed == original
