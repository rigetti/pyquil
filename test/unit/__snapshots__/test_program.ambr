# serializer version: 1
# name: test_with_loop
  '''
  DECLARE ro BIT[1]
  DECLARE shot_count INTEGER[1]
  DEFFRAME 0 "rx":
  	HARDWARE-OBJECT: "hardware"
  DEFWAVEFORM custom:
  	1, 2
  DEFCAL I 0:
  	DELAY 0 1
  MOVE shot_count[0] 100
  LABEL @start-loop
  MEASURE q ro[0]
  JUMP-UNLESS @end-reset ro[0]
  X q
  LABEL @end-reset
  I 0
  SUB shot_count[0] 1
  JUMP-UNLESS @end-loop shot_count[0]
  JUMP @start-loop
  LABEL @end-loop
  
  '''
# ---
